### PR TITLE
IBX-847: Request Lifecycle

### PR DESCRIPTION
**Pull request by adriendupuis**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [IBX-847](https://issues.ibexa.co/browse/IBX-847)
| Versions      | 3.3

Add a guide page about Handling a Request and sending back a Response with schemas.

### CODE REVIEW

File: docs/guide/request_lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+The schemas start with a Request object entering Symfony and Ibexa DXP. No REST or GraphQL HTTP Request here, only regular request from a browser.
+
+
+
+## From Request to Response: Flowcharts
+
+### Concept Flowchart
+
+![](img/request_lifecycle_concept.png)
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:32]:
```suggestion
![Simplified request lifecycle flowchart](img/request_lifecycle_concept.png)
```
We always add the alt text to images

File: docs/guide/request_lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+*Notice about Permission Control*: Another `kernel.request` event listener is the EzSystems\EzPlatformAdminUi\EventListener\RequestListener (priority 13): When a route got a `siteaccess_group_whitelist` parameter, this listener check that the current siteaccess is in one of the listed groups. For example, the Admin UI set an early protection of its route by passing them a `siteaccess_group_whitelist` containing only the `admin_group`.
+[TODO: https://doc.ibexa.co/en/latest/guide/permissions/#permissions-for-routes]
+
+Now, the Request know its controller, the HttpKernel dispatch the `kernel.controller` event.
+
+
+
+## Kernel's Controller Event
+
+(see `bin/console debug:event-dispatcher kernel.controller;`)
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:90]: I'm wondering in what situation a user may need this information? Is there a use case for it?

Anyway, if there is, I'd suggest something like:

```suggestion
To view all listeners that listen to this event, run `php bin/console debug:event-dispatcher kernel.controller`.
```

File: docs/guide/request_lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+If the HTTP request is to be treated by Ibexa DXP, it goes to the [Symfony Front Controller](https://symfony.com/doc/current/configuration/front_controllers_and_kernel.html#the-front-controller) public/index.php
+
+The Front Controller transform the HTTP request into a PHP object `Request` and pass it to Symfony's Kernel to obtain a `Response` object that will be sent back as an HTTP response.
+
+The schemas start with a Request object entering Symfony and Ibexa DXP. No REST or GraphQL HTTP Request here, only regular request from a browser.
+
+
+
+## From Request to Response: Flowcharts
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:28]: We use sentence case in headings (only first word, proper names and domain terms from a fixed listed are capitalized):

```suggestion
## From request to response: flowcharts
```

File: docs/guide/request_lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+- then dispatch the `ezpublish.siteaccess` event (a.k.a MVCEvents::SITEACCESS) (see `bin/console debug:event-dispatcher ezpublish.siteaccess;`).
+
+The SiteAccessListener subscribes to this `ezpublish.siteaccess` event with top priority. The SiteAccessListener add the attribute `semanticPathinfo`, the path without siteaccess indications ([URIElement](https://doc.ibexa.co/en/latest/guide/multisite/siteaccess_matching/#urielement), [URIText](https://doc.ibexa.co/en/latest/guide/multisite/siteaccess_matching/#uritext) or [Map\URI](https://doc.ibexa.co/en/latest/guide/multisite/siteaccess_matching/#mapuri) implementing the URILexer interface), to the Request.
+
+
+### Routing
+
+Lately, also listening the `kernel.request` event, the Symfony\Component\HttpKernel\EventListener\RouterListener (priority 32) calls the eZ\Publish\Core\MVC\Symfony\Routing\ChainRouter::matchRequest and add its returned parameters to the Request.
+
+ChainRouter:
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:63]: these deserve a heading, I think
```suggestion
#### ChainRouter

```

File: docs/guide/request_lifecycle.md (hunk 2 above)
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md:90]: @DominikaK I didn't know if this level of detail was relevant. As a developer, I found observing the result of this command instructive. But it also sounds like asking to the reader to document himself. I'll probably resume all those invitations to execute a command to only one `!!!tip` for people who want to know more about events and services cited in the article.

Comment (DominikaK) [reply docs/guide/request_lifecycle.md:90]: Yep, one tip sounds like a good idea.

File: docs/guide/request_lifecycle.md (hunk 1 above)
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done in 7342872c31219f64aa985c153a164765a55a5ac4

File: docs/guide/request_lifecycle.md (hunk 4 above)
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done. But only for the ChainRouter. Maybe Default, UrlWildcardRouter and UrlAliasRouter deserve a heading too.

File: docs/guide/request_lifecycle.md (hunk 3 above)
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md:21]: @DominikaK 
Before fixing this usage, I must fix the way I use PHP class names. Here, Request and Response refers to PHP class names and should be between backquotes. About class names usage, I have a difficulty with services. I often indicate both there class and their code name. For example “DefaultRouter (`router.default`)”. Maybe I could only use the code names as they're used in config files or when using `bin/console debug:container`.

File: docs/guide/request_lifecycle.md (hunk 4 above)
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done

File: docs/guide/request_lifecycle.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+
+
+## Kernel's controller event
+
+### View building and matching
+
+The HttpKernel just dispatched the `kernel.controller` event.
+
+Listening to `kernel.controller`, the `ViewControllerListener` (`ezpublish.view_controller_listener`) (priority 10) checks if the `_controller` Request attribute is associated to a `ViewBuilder` (a service tagged `ibexa.view_builder`) in the `ViewBuilderRegistry` (`ezpublish.view_builder.registry`).
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:107]:
```suggestion
Listening to `kernel.controller`, the `ViewControllerListener` (`ezpublish.view_controller_listener`) (priority 10) checks if the `_controller` request attribute is associated with a `ViewBuilder` (a service tagged `ibexa.view_builder`) in the `ViewBuilderRegistry` (`ezpublish.view_builder.registry`).
```

File: docs/guide/request_lifecycle.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+The `ContentViewBuilder` (`ezpublish.view_builder.content`) matches on controller starting with `ez_content:` (see `eZ\Publish\Core\MVC\Symfony\View\Builder\ContentViewBuilder::matches`).
+The `ContentViewBuilder` builds a `ContentView`.
+
+First, the `ContentViewBuilder` will load the `Location` and the `Content`, and, add them to the `ContentView` object.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:111]:
```suggestion
First, the `ContentViewBuilder` loads the `Location` and the `Content`, and adds them to the `ContentView` object.
```

File: docs/guide/request_lifecycle.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+**Notice about Permission Control**: `content/read` and/or `content/view_embed` permissions are controlled during this `ContentView` building.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:113]:
```suggestion
!!! caution "Permission control"

    `content/read` and/or `content/view_embed` permissions are controlled during this `ContentView` building.
```

File: docs/guide/request_lifecycle.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+Then, the `ContentViewBuilder` passes the `ContentView` to its `View\Configurator` (`ezpublish.view.configurator`).
+It's implemented by the `View\Configurator\ViewProvider` and its `View\Provider\Registry`, this registry receives the services tagged `ezpublish.view_provider` thanks to the `ViewProviderPass`.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:116]:
```suggestion
It's implemented by the `View\Configurator\ViewProvider` and its `View\Provider\Registry`. This registry receives the services tagged `ezpublish.view_provider` thanks to the `ViewProviderPass`.
```

File: docs/guide/request_lifecycle.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+Among the view providers, the services using the `eZ\Bundle\EzPublishCoreBundle\View\Provider\Configured` have an implementation of the `MatcherFactoryInterface` (`ezpublish.content_view.matcher_factory`).
+Through service decoration and class inheritance, the `ClassNameMatcherFactory` will be responsible for the [View Matching](content_rendering/templates/template_configuration/#view-rules-and-matching).
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:118]:
```suggestion
Through service decoration and class inheritance, the `ClassNameMatcherFactory` is responsible for the [view matching](content_rendering/templates/template_configuration/#view-rules-and-matching).
```

File: docs/guide/request_lifecycle.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+The `View\Configurator\ViewProvider` will use the matched view rule to add possible **`templateIdentifier`** and **`controllerReference`** to the `ContentView` object.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:119]:
```suggestion
The `View\Configurator\ViewProvider` uses the matched view rule to add possible **`templateIdentifier`** and **`controllerReference`** to the `ContentView` object.
```

File: docs/guide/request_lifecycle.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+The `ViewControllerListener` adds the ContentView to the `Request` as **`view`** attribute.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:121]:
```suggestion
The `ViewControllerListener` adds the ContentView to the `Request` as the **`view`** attribute.
```

File: docs/guide/request_lifecycle.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+The `ViewControllerListener` eventually updates the `Request` `_controller` attribute with the `ContentView`'s `controllerReference`.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:122]:
```suggestion
The `ViewControllerListener` eventually updates the request's `_controller` attribute with the `ContentView`'s `controllerReference`.
```

File: docs/guide/request_lifecycle.md [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+The `HttpKernel` then dispatches a `kernel.controller_arguments` (a.k.a. `KernelEvents::CONTROLLER_ARGUMENTS`) but nothing from Ibexa DXP is listening to it.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:124]:
```suggestion
The `HttpKernel` then dispatches a `kernel.controller_arguments` (`KernelEvents::CONTROLLER_ARGUMENTS`) but nothing from [[= product_name =]] listens to it.
```

File: docs/guide/request_lifecycle.md [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+
+
+## Controller execution
+
+The `HttpKernel` extracts from the Request the controller and the arguments to pass to the controller. [Argument resolvers](https://symfony.com/doc/current/controller/argument_value_resolver.html) way is similar to autowiring.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:130]:
```suggestion
The `HttpKernel` extracts from the request the controller and the arguments to pass to the controller. [Argument resolvers](https://symfony.com/doc/current/controller/argument_value_resolver.html) work in a way similar to autowiring.
```

File: docs/guide/request_lifecycle.md [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+The `HttpKernel` executes the controller with those arguments.
+
+As a reminder, the controller and its argument can be:
+
+- A controller set by a route matching and the `Request` as its argument.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:135]:
```suggestion
- A controller set by the matched route and the request as its argument.
```

File: docs/guide/request_lifecycle.md [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+- The default `ez_content:viewAction` controller and a `ContentView` as its argument.
+- A [custom controller](content_rendering/queries_and_controllers/controllers/) set by the matched view rule and a `View` or the `Request` as its argument (most likely a `ContentView` but there is no restriction). **Notice about Permission Control**: See [Permissions for custom controller](https://doc.ibexa.co/en/latest/guide/permissions/#permissions-for-custom-controllers).
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:137]:
```suggestion
- A [custom controller](content_rendering/queries_and_controllers/controllers.md) set by the matched view rule and a `View` or the request as its argument (most likely a `ContentView` but there is no restriction).

!!! caution "Permission Control"

    See [Permissions for custom controller](permissions.md#permissions-for-custom-controllers).
```
Hint: internal links should ideally be indicated used with `.md` instead of just `/`, as otherwise the automatic link checker does not work.

File: docs/guide/request_lifecycle.md [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+
+
+
+## Kernel's view event and `ContentView` rendering
+
+If the controller returns something else than a `Response`, the `HttpKernel` dispatches a `kernel.view` event (`KernelEvents::VIEW`).
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:143]:
```suggestion
If the controller returns something other than `Response`, the `HttpKernel` dispatches a `kernel.view` event (`KernelEvents::VIEW`).
```

File: docs/guide/request_lifecycle.md [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+In the case of a URL Alias, the controller most likely returns a ContentView.
+The `ViewRendererListener` (`ezpublish.view.renderer_listener`) will use the `ContentView` and the `TemplateRenderer` (`ezpublish.view.template_renderer`) to get the content of the `Response` and attach this new `Response` to the event.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:145]:
```suggestion
The `ViewRendererListener` (`ezpublish.view.renderer_listener`) uses the `ContentView` and the `TemplateRenderer` (`ezpublish.view.template_renderer`) to get the content of the `Response` and attach this new `Response` to the event.
```

File: docs/guide/request_lifecycle.md [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+The `HttpKernel` retrieve the Response attached to the event and continue.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:146]:
```suggestion
The `HttpKernel` retrieves the response attached to the event and continues.
```

File: docs/guide/request_lifecycle.md [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+
+
+
+## Kernel's response event and `Response` sending
+
+The `HttpKernel` send a `kernel.response` event (`KernelEvents::RESPONSE`). For example, if HTTP Cache is used, `Response`'s headers may be enhanced.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:152]:
```suggestion
The `HttpKernel` sends a `kernel.response` event (`KernelEvents::RESPONSE`). For example, if HTTP cache is used, response's headers may be enhanced.
```

File: docs/guide/request_lifecycle.md [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+
+The `HttpKernel` send a `kernel.finish_request` event (`KernelEvents::FINISH_REQUEST`). The `VerifyUserPoliciesRequestListener` (`siso_core.verify_user_policies_request_listener`) (priority 100) is filtering route on its policy configuration. **Notice about Permission Control**: See [Permissions for routes](permissions/#permissions-for-routes).
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:154]:
```suggestion
The `HttpKernel` sends a `kernel.finish_request` event (`KernelEvents::FINISH_REQUEST`). The `VerifyUserPoliciesRequestListener` (`siso_core.verify_user_policies_request_listener`) (priority 100) filters routes on its policy configuration.

 !!! note "Permission control
 
     See [Permissions for routes](permissions.md#permissions-for-routes).
```

File: docs/guide/request_lifecycle.md [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+
+Finally, the `HttpKernel` send the `Response`.
+
+If an exception occurs during this chain of events, the `HttpKernel` send a `kernel.exception` and try to obtain a `Response` from its listeners.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:158]:
```suggestion
If an exception occurs during this chain of events, the `HttpKernel` sends a `kernel.exception` and tries to get a `Response` from its listeners.
```

File: docs/guide/request_lifecycle.md [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+
+The `HttpKernel` send a last `kernel.terminate` event (`KernelEvents::TERMINATE`). For example, the `BackgroundIndexingTerminateListener` (`ezpublish.search.background_indexer`) (priority 0) remove from the `SearchService` index possible contents existing in the index but not in the database.
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:160]:
```suggestion
The `HttpKernel` sends the last `kernel.terminate` event (`KernelEvents::TERMINATE`). For example, the `BackgroundIndexingTerminateListener` (`ezpublish.search.background_indexer`) (priority 0) removes from the `SearchService` index possible content existing in the index but not in the database.
```

**Review by DominikaK — CHANGES_REQUESTED** [review]

File: docs/guide/request_lifecycle.md [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+
+![Simplified request lifecycle flowchart](img/request_lifecycle_concept.png)
+
+The next chart will detail how this is implemented using Symfony HTTP kernel listeners.
+
+### Kernel events flowchart
+
+The following chart shows events, listeners and attributes added to the request or its wrapping event object.
+
+<img src="../img/request_lifecycle_events.png" width="741" height="2061" style="max-height: none;" alt="Detailed request lifecycle flowchart organised around kernel events" />
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:37]: We could actually use markdown here if we make use of one of the plugins available for mkdocs.
This would require adding to mkdocs.yml:
```
markdown_extensions:
  - attr_list
 ```

Then the following should work:
```
![Detailed request lifecycle flowchart organised around kernel events](img/request_lifecycle_events.png){: width="741" height="2061" style="max-height: none;"}
```

File: docs/guide/request_lifecycle.md [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+
+## HTTP request entering
+
+When entering the architecture, the HTTP request can be handled by several component like a firewall, a load balancer, a reverse-proxy, etc. before arriving on the web server itself.
+
+For an overview of what happens on reverse proxy like Varnish or Fastly, see [Context-aware HTTP cache / Request lifecycle](cache/context_aware_cache/#request-lifecycle).
+
+When arriving at a Web server, the request is filtered by Apache Virtual Host, Nginx Server Blocks or equivalent. This mainly (and shortly) separates requests of static resources from requests to the application.
+
+As Ibexa DXP is a Symfony application, everything starts like in Symfony &mdash; see [Symfony and HTTP Fundamentals](https://symfony.com/doc/current/introduction/http_fundamentals.html).
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done in 59ab776e48e695cb9355bfe87f47a6011f40365a

File: docs/guide/request_lifecycle.md (hunk 24 above)
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done in d4376faa5d676e4ef718ec054671467a3a3c0af7

File: docs/guide/request_lifecycle.md [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+
+When arriving at a Web server, the request is filtered by Apache Virtual Host, Nginx Server Blocks or equivalent. This mainly (and shortly) separates requests of static resources from requests to the application.
+
+As Ibexa DXP is a Symfony application, everything starts like in Symfony, see [Symfony and HTTP Fundamentals](https://symfony.com/doc/current/introduction/http_fundamentals.html).
+
+If the HTTP request is to be treated by Ibexa DXP, it goes to the [Symfony Front Controller](https://symfony.com/doc/current/configuration/front_controllers_and_kernel.html#the-front-controller) public/index.php
+
+The Front Controller transform the HTTP request into a PHP [`Request` object](https://symfony.com/doc/current/introduction/http_fundamentals.html#symfony-request-object) and pass it to Symfony's Kernel to obtain a [`Response` object](https://symfony.com/doc/current/introduction/http_fundamentals.html#symfony-response-object) that will be transform and sent back as an HTTP response.
+
+The schemas start with a `Request` object entering Symfony and Ibexa DXP, only regular request from a browser. No ESI, no REST nor GraphQL request.
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done in 5beda4b68204fbf45c06ae5db754267a4b8c036e

File: docs/guide/request_lifecycle.md [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+The `SiteAccessListener` (`ezpublish.siteaccess_listener`) subscribes to this `ezpublish.siteaccess` event with top priority (priority 255). The `SiteAccessListener` add the attribute **`semanticPathinfo`**, the path without siteaccess indications ([`URIElement`](multisite/siteaccess_matching/#urielement), [`URIText`](multisite/siteaccess_matching/#uritext) or [`Map\URI`](multisite/siteaccess_matching/#mapuri) implementing the `URILexer` interface), to the Request.
+
+
+### Routing
+
+Lately, also listening the `kernel.request` event, the `Symfony\Component\HttpKernel\EventListener\RouterListener` (`router_listener`)(priority 32) calls the `eZ\Publish\Core\MVC\Symfony\Routing\ChainRouter::matchRequest` and add its returned parameters to the `Request`.
+
+#### ChainRouter
+
+The [`ChainRouter`](https://symfony.com/doc/current/cmf/components/routing/chain.html) is a Symfony Content Management Framework (CMF) component. Ibexa DXP makes it a service named `ezpublish.chain_router`.
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done

File: docs/guide/request_lifecycle.md [hunk 28]
```diff
@@ -0,0 +1,221 @@
...
+It has a collection of prioritized routers to crawl in until finding one matching the Request to define what to do next.
+The `ChainRouter` router collection is build by the ChainRoutingPass collecting the services tagged `router`.
+The `DefaultRouter` is always added to the collection with top priority (priority 255).
+
+`DefaultRouter` (`router.default`):
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md:97]: Done in cad2354868f76fd4d655b6fe9048fb652fd3ba49

File: docs/guide/request_lifecycle.md [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+
+As Ibexa DXP is a Symfony application, everything starts like in Symfony, see [Symfony and HTTP Fundamentals](https://symfony.com/doc/current/introduction/http_fundamentals.html).
+
+If the HTTP request is to be treated by Ibexa DXP, it goes to the [Symfony Front Controller](https://symfony.com/doc/current/configuration/front_controllers_and_kernel.html#the-front-controller) public/index.php
+
+The Front Controller transform the HTTP request into a PHP [`Request` object](https://symfony.com/doc/current/introduction/http_fundamentals.html#symfony-request-object) and pass it to Symfony's Kernel to obtain a [`Response` object](https://symfony.com/doc/current/introduction/http_fundamentals.html#symfony-response-object) that will be transform and sent back as an HTTP response.
+
+The schemas start with a `Request` object entering Symfony and Ibexa DXP, only regular request from a browser. No ESI, no REST nor GraphQL request.
+
+
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md]: Done in 81a50f5ee62beda2e2951ff4b32f8af7cfe15a3e?

File: docs/guide/request_lifecycle.md [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+This schema is detailed below event by event. 
+
+!!! tip
+
+    - To list all listeners that listen to an event, run `php bin/console debug:event-dispatcher <event.name>`;<br/>For example `bin/console debug:event-dispatcher kernel.request`
+    - To view details of a service (including class, arguments and tags), run `bin/console debug:container --show-arguments <service.name>`;<br/>For example `bin/console debug:container --show-arguments ezpublish.siteaccess_match_listener`
+    - To list all services with a specific tag, run `bin/console debug:container --tag=<tag>>`;<br/>For example, `bin/console debug:container --tag=router`
+
+
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md:58]: Done in 81a50f5ee62beda2e2951ff4b32f8af7cfe15a3e?

File: docs/guide/request_lifecycle.md [hunk 31]
```diff
@@ -0,0 +1,223 @@
...
+
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/guide/request_lifecycle.md:59]: Done in 81a50f5ee62beda2e2951ff4b32f8af7cfe15a3e?

**Comment by adriendupuis** [comment]
@adriendupuis TODO: Update internal links to .md

File: docs/guide/request_lifecycle.md [hunk 32]
```diff
@@ -0,0 +1,253 @@
...
+### `UrlAliasRouter`
+
+`UrlAliasRouter` (`ezpublish.urlalias_router`):
+This router uses the `UrlAliasService` to associate the `semanticPathinfo` to a Location.
+If it finds a Location, the request receives the attributes **`locationId`** and **`contentId`**, **`viewType`** is set to `full`, and the **`_controller`** is set to `ez_content:viewAction` for now.
+
+The `locale_listener` (priority 16) sets the request's **`_locale`** attribute.
+
+!!! caution "Permission control"
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:115]:
```suggestion
!!! note "Permission control"
```
Actually when I think about it, these don't need to be cautions, just notes (cautions are reserved for dangerous/critical notes)

File: docs/guide/request_lifecycle.md [hunk 33]
```diff
@@ -0,0 +1,253 @@
...
+     See [Permissions for routes](permissions.md#permissions-for-routes).
+
+Finally, the `HttpKernel` send the response.
+
+If an exception occurs during this chain of events, the `HttpKernel` sends a `kernel.exception` and tries to get a `Response` from its listeners.
+
+The `HttpKernel` sends the last `kernel.terminate` event (`KernelEvents::TERMINATE`). For example, the `BackgroundIndexingTerminateListener` (`ezpublish.search.background_indexer`) (priority 0) removes from the `SearchService` index possible content existing in the index but not in the database.
+
+
+## Summaries
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:193]:
```suggestion
## Summary
```

File: docs/guide/request_lifecycle.md [hunk 34]
```diff
@@ -0,0 +1,253 @@
...
+        - `ezpublish.view.template_renderer`
+* event=`kernel.response`
+* event=`kernel.finish_request`
+    - 100:`siso_core.verify_user_policies_request_listener`
+* event=`kernel.terminate`
+    - 0:`ezpublish.search.background_indexer`
+
+### Examples request attributes timeline
+
+|  Event                  |  Service                              |  Request attribute  |  Example #1   |  Example #2     |
```
Comment (DominikaK) [inline docs/guide/request_lifecycle.md:228]: The table is too wide and render in an unreadable way on mid-sized screens. Would it hurt to keep only one example column? Or redo it in some other way.

**Review by DominikaK — CHANGES_REQUESTED** [review]

File: docs/guide/request_lifecycle.md [hunk 35]
```diff
@@ -0,0 +1,252 @@
+# Request lifecycle: from request to response
+
+
+## Beginning of HTTP request
+
+When entering the architecture, the HTTP request can be handled by several component such as a firewall, a load balancer, or a reverse proxy before arriving on the web server itself.
```
Comment (Steveb-p) [inline docs/guide/request_lifecycle.md:6]:
```suggestion
When entering the server infrastructure, the HTTP request can be handled by several component such as a firewall, a load balancer, or a reverse proxy before arriving on the web server itself.
```

File: docs/guide/request_lifecycle.md [hunk 36]
```diff
@@ -0,0 +1,253 @@
+# Request lifecycle: from request to response
+
+
+## Beginning of HTTP request
+
+When entering the architecture, the HTTP request can be handled by several component such as a firewall, a load balancer, or a reverse-proxy before arriving on the web server itself.
+
+For an overview of what happens on a reverse proxy like Varnish or Fastly, see [Context-aware HTTP cache / Request lifecycle](cache/context_aware_cache/#request-lifecycle).
+
+When arriving at a web server, the request is filtered by Apache Virtual Host, Nginx Server Blocks or equivalent. This mainly (and shortly) separates requests of static resources from requests to the application.
```
[the remark this answers is not in the document]
Comment (Steveb-p) [reply docs/guide/request_lifecycle.md:10]: > Given that we only support Apache or nginx, does mentioning "equivalent" make sense here?

In general, any HTTP server that passes the request onto PHP interpreter can be used. In a sense web-server acts as a reverse-proxy here, passing the request to a "third-party" (PHP) and presenting the response as it's own.

> There, requests of static resources are quickly separated from requests to the application.

The more important part here is that static resources are served quickly. We're talking about 1000 requests served by PHP vs. 100k statically served files (the order of magnitude differs, but that's the change you could expect).

However, this also means that any security checks that would be handled by PHP are omitted in this case, so any content that should be reachable only by application users should not be handled in this manner.

File: docs/guide/request_lifecycle.md [hunk 37]
```diff
@@ -0,0 +1,252 @@
...
+
+## Lifecycle flowcharts
+
+### Concept flowchart
+
+The chart below introduces the logic of the request treatment.
+
+![Simplified request lifecycle flowchart](img/request_lifecycle_concept.png)
+
+
```
Comment (Steveb-p) [inline docs/guide/request_lifecycle.md:29]: Unnecessary whitespace.
```suggestion
```

File: docs/guide/request_lifecycle.md [hunk 38]
```diff
@@ -0,0 +1,252 @@
...
+![Detailed request lifecycle flowchart organised around kernel events](img/request_lifecycle_events.png){: width="741" height="2061" style="max-height: none;"}
+
+This schema is described below event by event.
+
+!!! tip
+
+    To list all listeners that listen to an event, run `php bin/console debug:event-dispatcher <event.name>`, for example:
+    
+    ```bash
+    bin/console debug:event-dispatcher kernel.request
```
Comment (Steveb-p) [inline docs/guide/request_lifecycle.md:43]:
```suggestion
    php bin/console debug:event-dispatcher kernel.request
```

File: docs/guide/request_lifecycle.md [hunk 39]
```diff
@@ -0,0 +1,252 @@
...
+    ```
+    
+    To view details of a service (including class, arguments and tags), run `bin/console debug:container --show-arguments <service.name>`, for example:
+    
+    ```bash
+    bin/console debug:container --show-arguments ezpublish.siteaccess_match_listener`
```
Comment (Steveb-p) [inline docs/guide/request_lifecycle.md:49]:
```suggestion
    php bin/console debug:container --show-arguments ezpublish.siteaccess_match_listener`
```

File: docs/guide/request_lifecycle.md [hunk 40]
```diff
@@ -0,0 +1,252 @@
...
+
+!!! tip
+
+    To list all listeners that listen to an event, run `php bin/console debug:event-dispatcher <event.name>`, for example:
+    
+    ```bash
+    bin/console debug:event-dispatcher kernel.request
+    ```
+    
+    To view details of a service (including class, arguments and tags), run `bin/console debug:container --show-arguments <service.name>`, for example:
```
Comment (Steveb-p) [inline docs/guide/request_lifecycle.md:46]:
```suggestion
    To view details of a service (including class, arguments and tags), run `php bin/console debug:container --show-arguments <service.name>`, for example:
```

File: docs/guide/request_lifecycle.md [hunk 41]
```diff
@@ -0,0 +1,252 @@
...
+    
+    ```bash
+    bin/console debug:container --show-arguments ezpublish.siteaccess_match_listener`
+    ```
+    
+    To list all services with a specific tag, run `bin/console debug:container --tag=<tag>`, for example:
```
Comment (Steveb-p) [inline docs/guide/request_lifecycle.md:52]:
```suggestion
    To list all services with a specific tag, run `php bin/console debug:container --tag=<tag>`, for example:
```

File: docs/guide/request_lifecycle.md [hunk 42]
```diff
@@ -0,0 +1,252 @@
...
+    
+    ```bash
+    bin/console debug:container --tag=router
```
Comment (Steveb-p) [inline docs/guide/request_lifecycle.md:55]:
```suggestion
    php bin/console debug:container --tag=router
```

File: docs/guide/request_lifecycle.md (hunk 38 above)
Comment (DominikaK) [reply docs/guide/request_lifecycle.md:43]: Agreed, that's the convention, I missed it.

**Review by DominikaK — APPROVED** [review]
Looking good now.

**Comment by adriendupuis** [comment]
- In 3.3: 8bdc81c65e534b441dde3cd6aaf711c3ef80f859
- In master: 9e87f731c8182fea7865a9a6b4a04e9088c29706
